### PR TITLE
Dev 29 - create-script-to-generate-crops-from-a-dataset

### PR DESCRIPTION
**Pull request by powl7**
### Description
Added a script that create a new dataset by cropping an existing dataset. The crops are created using a sliding window (same as the sliding window used when testing).

### How to Test/Run?
python src/generate_cropped_dataset.py
(To generate a different cropped dataset, the part under `if __name__ == '__main__':` has to be changed.)
I ran the script on Dana.

### Something missing?
I have not yet tested if the dataset works as expected. I created a trello ticket for this.

### CODE REVIEW

File: src/generate_cropped_dataset.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+    """
+
+    paths = []
+    directory = os.path.expanduser(directory)
+
+    path_imgs = os.path.join(directory, "data")
+    path_gts = os.path.join(directory, "gt")
+
+    if not (os.path.isdir(path_imgs) or os.path.isdir(path_gts)):
```
Comment (lvoegtlin) [inline src/generate_cropped_dataset.py]: Why not using the Patch object as you already use it in other positions

File: src/generate_cropped_dataset.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        logging.error("folder data or gt not found in " + str(directory))
+
+    for img_name, gt_name in zip(sorted(os.listdir(path_imgs)), sorted(os.listdir(path_gts))):
+        assert has_extension(img_name, IMG_EXTENSIONS) == has_extension(gt_name, IMG_EXTENSIONS), \
+            'get_gt_data_paths(): image file aligned with non-image file'
+
+        if has_extension(img_name, IMG_EXTENSIONS) and has_extension(gt_name, IMG_EXTENSIONS):
+            assert os.path.splitext(img_name)[0] == os.path.splitext(gt_name)[0], \
+                'get_gt_data_paths(): mismatch between data filename and gt filename'
+            paths.append((os.path.join(path_imgs, img_name), os.path.join(path_gts, gt_name)))
```
Comment (lvoegtlin) [inline src/generate_cropped_dataset.py]: same here with the Path

File: src/generate_cropped_dataset.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        img_names_sizes = []  # list of tuples -> (gt_img_name, img_size (H, W))
+        num_horiz_crops = []
+        num_vert_crops = []
+
+        for img_path, gt_path in self.img_paths:
+            data_img = pil_loader(img_path)
+            gt_img = pil_loader(gt_path)
+            # Ensure that data and gt image are of the same size
+            assert gt_img.size == data_img.size
+            img_names_sizes.append((os.path.basename(gt_path), data_img.size[::-1]))
```
Comment (lvoegtlin) [inline src/generate_cropped_dataset.py]: also herewith Path

File: src/generate_cropped_dataset.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    dataset_generator = CroppedDatasetGenerator(
+        input_path=Path('/dataset/DIVA-HisDB/segmentation/CB55-10-segmentation'),
+        output_path=Path('/data/usl_experiments/semantic_segmentation/datasets_cropped/CB55-10-segmentation'),
+        crop_size_train=300,
+        crop_size_val=300,
+        crop_size_test=256,
+        overlap=0.5,
+        leading_zeros_length=4,
+        override_existing=True)
+    dataset_generator.write_crops()
```
Comment (lvoegtlin) [inline src/generate_cropped_dataset.py]: Perhaps we could create a cli out of it (like we are doing in the other scripts)

File: src/generate_cropped_dataset.py (hunk 1 above)
Comment (powl7) [reply src/generate_cropped_dataset.py]: legacy implementation

File: src/generate_cropped_dataset.py (hunk 2 above)
Comment (powl7) [reply src/generate_cropped_dataset.py]: legacy implementation

File: src/generate_cropped_dataset.py (hunk 3 above)
Comment (powl7) [reply src/generate_cropped_dataset.py]: legacy implementation

File: src/generate_cropped_dataset.py (hunk 4 above)
Comment (powl7) [reply src/generate_cropped_dataset.py]: Good idea.

**Comment by lvoegtlin** [comment]
added a CLI and replaced the old os calls with path calls

**Comment by lvoegtlin** [comment]
[Create script to generate crops from a dataset](https://trello.com/c/TkIYxKVL/29-create-script-to-generate-crops-from-a-dataset)

**Pull request merged**